### ISSUE TITLE
Synthesizing Tabular Data using Generative Adversarial Networks

### ISSUE THREAD

**Issue by ghuijo** [yskim0/One-Day-One-Paper#84]
* CTGAN 이전에 나온 TGAN 모델.

## Abstract
* Tabular GAN (TGAN) 제안.
* medical이나 educational 기록과 같은 tabular data를 생성할 수 있음.
* discrete하거나 continuous한 variables 모두 생성 가능함.

## Introduction
* synthetic data 어디에 쓰냐?
  * 외부인이 data에 access하는 것을 막기 위해 등등.
* 기존 generative model들은 distribution functions의 type이 제한적이었음.
* 본 논문에서는 synthetic data generator를 tabular data를 위한 GAN 기반으로 develop함.
* mixed variable types (multinomial/discrete&continuous)의 tabular data 생성에 집중함.
* **LSTM with attention을 comlumn별로 data를 생성하기 위해 사용함.**

## Related Work
* 기존 GAN 모델들과 synthetic data generation 위한 모델들.
* (RCGAN이랑 medGAN 논문 읽어보자)

## GANs for tabular data
* table에는 다양한 유형의 data가 있을 수 있고(numerical, categorical 등), variables의 분포 모양도 다양함(multimodal, long tail 등).
* 간단한 notation 정리 먼저
  * table T / continuous random variables는 C아래첨자로 / discrete random variables는 D아래첨자로 / variable들의 unknown joint distribution은 P(C, D).
* model 방향 2가지
  * A machine learning model learned using Tsynth can achieve a similar accuracy on a real test table Ttest (usually set aside at the beginning), as would a model learned using the data from table T.
  * Mutual information: The mutual information between an arbitrary pair of variables i, j in T and Tsynth is similar.
* reversible transformation을 적용함.
  * numerical variable을 (-1, 1) 범위 & multinomial distribution으로 변환함.
  * discrete variable을 multinomial distribution으로 변환함.
* mode-specific normalization
  * continuous variable의 mode 수를 estimate하기 위해 Gaussian kernel density estimation 사용함.
  * 단순히 [-1, 1] 범위로 normalizing하면 gradient가 -1, 1과 가까운 mode일 때 saturate할 수 있음.
  * 그래서 Gaussian Mixture model (GMM)을 써서 numeriacal variable을 cluster함.
<image src="https://user-images.githubusercontent.com/67627471/130081124-1538ee0c-41f3-4aaf-84c4-9668307a9b65.png" width="70%" height="70%" />

  * uni-modal과 multi-modal features 모두 m=5 Gaussian distributions로 cluster함.
* Smoothing for categorical variables (how to make the model differentiable)
  * generate the probability distribution directly using softmax.
  * --> 그러기 위해서, convert categorical variables to one-hot-encoding representation & add noise to binary variables.
<image src="https://user-images.githubusercontent.com/67627471/130081745-a5e9d511-3f74-4599-a948-3b77220908e1.png" width="70%" height="70%" />

* 전처리 후, n_c+n_d의 columns가진 테이블 T를 v_1:nc,j, u_1:nc,j, d_1:nd, j로 변환함. (벡터로 변환했다는 뜻..)
* TGAN 대략적인 구조
<image src="https://user-images.githubusercontent.com/67627471/130082251-aa449bf7-28bd-431b-b054-13f489fd5800.png" width="60%" height="60%" />

* LSTM을 generator에 쓰고, MLP를 discriminator에 씀.
#### Generator
* 일단, generate the value scalar v_i -> generate the cluster vector u_i -> categorical feature in 1 step as a probability distribution over all possible labels.
* attention weight vector를 학습함.
<image src="https://user-images.githubusercontent.com/67627471/130082780-da5341b7-c105-4006-a84b-d468f579a6f1.png" width="20%" height="20%" />

* LSTM의 output은 h_t임.
<image src="https://user-images.githubusercontent.com/67627471/130082916-f14d202b-3c10-4dec-8fe6-22cc8f152656.png" width="70%" height="70%" />

#### Discriminator
* l-layer fc nn을 discriminator로 씀.
<image src="https://user-images.githubusercontent.com/67627471/130083055-6c59360e-dea3-4678-a0ae-34f06c9f9b62.png" width="70%" height="70%" />

#### Loss function
* generator 학습용
<image src="https://user-images.githubusercontent.com/67627471/130083193-166f5912-261a-497a-a0b3-604eae02eb7a.png" width="50%" height="50%" />

* discriminator 학습용
<image src="https://user-images.githubusercontent.com/67627471/130083290-bd714e1b-461c-48a8-a1b8-59903ff3dc14.png" width="60%" height="60%" />

## Evaluation Setup
* first train a TGAN data synthesizer using the real training data T and generate a synthetic training dataset Tsynth.
* then train machine learning models on both the real and synthetic datasets.
<image src="https://user-images.githubusercontent.com/67627471/130083456-7f9b50f8-1b7b-48ca-a2a0-795ec6875056.png" width="50%" height="50%" />

## Results
* 일단 사용한 dataset 정보.
<image src="https://user-images.githubusercontent.com/67627471/130083600-de29ac20-e3fb-4ba3-8ff5-9a272485382c.png" width="70%" height="70%" />

* Census dataset을 학습한 모델들의 Macro-F1 evaluation.
<image src="https://user-images.githubusercontent.com/67627471/130083685-aceb4701-19eb-4022-bc86-8b8a579f2692.png" width="70%" height="70%" />

* real 데이터셋과 synthetic training set을 학습한 모델들의 정확도.
<image src="https://user-images.githubusercontent.com/67627471/130083766-f59d4c40-6554-4749-bf38-b16b7fef1a38.png" width="70%" height="70%" />

##
* CTGAN 읽기 전에 먼저 읽었어야 했는데..
